sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -235,6 +261,11 @@ object HiveTypeCoercion {
           && left.output.length == right.output.length && !s.resolved =>
         val (newLeft, newRight) = widenOutputTypes(s.nodeName, left, right)
         s.makeCopy(Array(newLeft, newRight))
+
+      case s: Union if s.childrenResolved &&
+        s.children.forall(_.output.length == s.children.head.output.length) && !s.resolved =>
```

Comment:
maybe indent 4 space here rather than 2 to make it clear this is not the same block as the next line